modules/internal/ChapelIO.chpl

```diff
@@ -17,36 +17,310 @@
+  other dimensions.
+* for a `domain` argument,  outputs the dimensions of the domain enclosed by
+  ``[`` and ``]``.
+* for a `range` argument, output the lower bound of the range followed by
+  ``..`` followed by the upper bound of the range.  If the stride of the range
+  is not one, the output is additionally followed by the word ``by`` followed
+  by the stride of the range.
+* for a tuples, outputs the components of the tuple in order delimited by ``(``
+  and ``)``, and separated by commas.
+* for a class, outputs the values within the fields of the class prefixed by
+  the name of the field and the character ``=``.  Each field is separated by a
+  comma.  The output is delimited by ``{`` and ``}``.
+* for a record, outputs the values within the fields of the class prefixed by
+  the name of the field and the character ``=``.  Each field is separated by a
+  comma.  The output is delimited by ``(`` and ``)``.
+
+Default ``read`` methods are created for all types for which a user-defined
+``read`` method is not provided.  The default ``read`` methods are defined to
+read in the output of the default ``write`` method.
+
+.. note::
+
+  Note that it is not currently possible to read and write circular
+  data structures with these mechanisms.
+
+ */
 module ChapelIO {
   use ChapelBase; // for uint().
   use SysBasic;
   // use IO; happens below once we need it.
   
+  // TODO -- this should probably be private
+  pragma "no doc"
   proc _isNilObject(val) {
     proc helper(o: object) return o == nil;
     proc helper(o)         return false;
     return helper(val);
   }
   
+  /*
+     An abstract base class type for classes that know how
+     to write values.
+   */
   class Writer {
+    /* returns `true` */
     proc writing param return true;
-    // if it's binary, we don't decorate class/record fields and values
+    /* returns `true` if this Writer is configured for binary I/O. */
     proc binary():bool { return false; }
-    // get other style elements
+    /* return other style elements.
+       subclasses should override this method */
+    pragma "no doc"
     proc styleElement(element:int):int { return 0; }
 
+    /* 
+       Return any saved error code.
+       Subclasses should override this method.
+     */
     proc error():syserr { return ENOERR; }
+    /*
+       Save an error code.
+       Subclasses should override this method.
+     */
     proc setError(e:syserr) { }
+    /* 
+       Clear any saved error code.
+       Subclasses should override this method.
+     */
     proc clearError() { }
+    /* Write a primitive type.
+       Subclasses should override this method.
+     */
     proc writePrimitive(x) {
       //compilerError("Generic Writer.writePrimitive called");
       halt("Generic Writer.writePrimitive called");
     }
+    /*
+       Write a sequence of bytes.
+       Subclasses should override this method.
+     */
+    pragma "no doc"
```

Comment:
I didn't document this one because the argument type is _ddata and that's not user-visible.